### PR TITLE
Adaptive average pooling 3d

### PR DESCRIPTION
**Pull request by ssnl**
Adding AdaptiveAvgPool3d as per request in #1988 .

Test plan:
./run_test.sh

### CODE REVIEW

**Comment by SsnL** [comment]
Hi there, this is my first time contributing. Let me know if I did anything wrong/suboptimally. :)

**Comment by soumith** [comment]
btw before you start on AdaptiveMaxPool3d, see the changes Greg made for the 2d class today: https://github.com/pytorch/pytorch/pull/2722

File: test/test_nn.py [hunk 1]
```diff
@@ -4024,6 +4024,24 @@ def add_test(test):
         desc='tuple',
     ),
     dict(
+        module_name='AdaptiveAvgPool3d',
+        constructor_args=((3),),
```
Comment (gchanan) [inline test/test_nn.py]: nit: the extra parens don't actually do anything here?

File: torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+#define START_IND(a,b,c) (int)floor((float)(a * c) / b)
+#define END_IND(a,b,c) (int)ceil((float)((a + 1) * c) / b)
+// #define START_IND(a,b,c) a * c / b
+// #define END_IND(a,b,c)  (a + 1) * c / b + ((a + 1) * c % b > 0)?1:0
+
+static void THNN_(VolumetricAdaptiveAveragePooling_updateOutput_frame)(
+          real *input_p,
+          real *output_p,
+          long nslices,
+          long isizeT,
```
Comment (gchanan) [inline torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c]: can you just use the same variable names / name formats as in SpatialAdaptiveAveragePooling?  Even if your names are better (they are), this creates unnecessary noise in review and maintenance.

File: torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+          long istrideH,
+          long strided)
+{
+  long k;
+#pragma omp parallel for private(k)
+  for (k = 0; k < nslices; k++)
+  {
+    /* loop over output */
+    long ti, i, j;
+    for(ti = 0; ti < osizeT; ti++)
```
Comment (gchanan) [inline torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c]: give "ti" a different name; from the name I'd expect it to loop over the time dimension of the input ("ot" doesn't seem unreasonable.

File: torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+  long k;
+#pragma omp parallel for private(k)
+  for (k = 0; k < nslices; k++)
+  {
+    real *gradInput_p_k = gradInput_p + k*isizeT*isizeW*isizeH;
+    real *gradOutput_p_k = gradOutput_p + k*osizeT*osizeW*osizeH;
+
+    /* calculate average */
+    long ti, i, j;
+    for(ti = 0; ti < osizeT; ti++)
```
Comment (gchanan) [inline torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c]: same here (both "isizeT", etc. and "ti"

**Review by gchanan — COMMENTED** [review]
I haven't looked at the cuda code yet (I'll do it tomorrow), but so far looks good!  Comments are minaly nits around minimizing the differences in the code with the Spatial versions.

File: torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c (hunk 2 above)
Comment (SsnL) [reply torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c]: I debated against myself on that for some time too. But I wanted to be consistent between the cpu and the cuda var names. I really didn't want to follow the spatial [cuda var names](https://github.com/pytorch/pytorch/blob/master/torch/lib/THCUNN/SpatialAdaptiveAveragePooling.cu#L21) which leads to `input_t` being a var rather than a type. What's your though on this?

Comment (soumith) [reply torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c]: if that's the case, just codemod the names in Spatial then to isize*.

File: torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+
+          int x, y, z;
+          for(z = z_start; z < z_end; z++)
+          {
+            for(y = y_start; y < y_end; y++)
+            {
+              for(x = x_start; x < x_end; x++)
+              {
+                /* update gradient */
+                gradInput_p_k[z*isizeH*isizeW + y*isizeW + x] += grad_delta;
```
Comment (SsnL) [inline torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c]: I'm suddenly confused by this line (and the corresponding line in spatial version). Why is this not atomic?

File: torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c (hunk 2 above)
Comment (SsnL) [reply torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c]: Should I do that in this PR or a separate one?

Comment (soumith) [reply torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c]: doing in this PR is fine.

File: torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c (hunk 5 above)
Comment (soumith) [reply torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c]: dont rule out bugs :)
It does indeed look like a race

Comment (ruotianluo) [reply torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c]: Because it's parallelized batch and channel dimension.

Comment (SsnL) [reply torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c]: Thanks for enlightening me, @ruotianluo !

File: torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c (hunk 2 above)
Comment (SsnL) [reply torch/lib/THNN/generic/VolumetricAdaptiveAveragePooling.c]: Will do tomorrow.

**Comment by SsnL** [comment]
Did massive renaming. I'd appreciate if someone can take a look. :)

File: torch/lib/THCUNN/generic/SpatialAdaptiveAveragePooling.cu [hunk 6]
```diff
@@ -20,59 +22,57 @@ void THNN_(SpatialAdaptiveAveragePooling_updateOutput)(
...
 
   if (input->nDimension == 3) {
-    long nInputCols = input->size[2];
-    long nInputRows = input->size[1];
-    long nInputPlane = input->size[0];
+    int sizeD  = input->size[0];
+    int isizeH = input->size[1];
+    int isizeW = input->size[2];
 
-    long istride_d = input->stride[0];
```
Comment (gchanan) [inline torch/lib/THCUNN/generic/SpatialAdaptiveAveragePooling.cu]: why change these from long to int?

File: torch/lib/THNN/generic/SpatialAdaptiveAveragePooling.c [hunk 7]
```diff
@@ -7,49 +7,51 @@
...
 // #define END_IND(a,b,c)  (a + 1) * c / b + ((a + 1) * c % b > 0)?1:0
 
+// 4d tensor B x D x H x W
+
 static void THNN_(SpatialAdaptiveAveragePooling_updateOutput_frame)(
           real *input_p,
           real *output_p,
-          long nslices,
-          long iwidth,
-          long iheight,
```
Comment (gchanan) [inline torch/lib/THNN/generic/SpatialAdaptiveAveragePooling.c]: same thing, why did these change?

**Review by gchanan — COMMENTED** [review]
1) Your changes now conflict, can you rebase and fix?
2) The spatial changes look correct, although I don't know why you changed some long types to int.  Can you explain? (I assume you meant to make them consistent, but don't know why you preferred int to long).
3) In general, it would be helpful if you made this easier to review.  For example, you could dedicate commits to just the spatial renames (no logic changes), then include logic changes in the next commit, then adding Volumetric files that are as close to the (changed) Spatial versions as possible, with clear commit messages about what each does.  I'm not asking you to change these now (except the above), just a note for the future.

**Comment by SsnL** [comment]
Thanks for your input @gchanan .

1 & 2. I preferred int because (a) in CUDA version, the kernel using them has int input parameters and (b) in CPU & CUDA version, the input indices computed from START_IND and END_IND are set to type int. So using long didn't seem necessary to me. Should I stick to using long (now int64_t in the master)?
The int64_t's are causing conflicts.  I will rebase & merge once this is settled.
3. Thanks for the tip! I will keep this in mind.

**Comment by gchanan** [comment]
> I preferred int because (a) in CUDA version, the kernel using them has int input parameters and (b) in CPU & CUDA version, the input indices computed from START_IND and END_IND are set to type int. So using long didn't seem necessary to me. Should I stick to using long (now int64_t in the master)?

Good question.  I discussed briefly with a couple of people and my best guess is this:
- In general, we only really care about performance on CUDA.  So the correct and performant thing to do there is to operate on ints if we can, otherwise operate on longs (you see this by searching for `canUse32BitIndexMath`.  Some of the implementations are lazy about this (or use some knowledge about the grid size) and always use ints.
- On cpu, we usually don't care that much and usually just use `long`.
- Some of your changes look incorrect.  For example, storing the strides as `int`s which are then applied as pointer arithmetic would previously work if everything but strides fit into `int`, but now they are broken.

In short, without a clear motivation beyond making everything the same type, I'm loathe to change the types.

**Comment by SsnL** [comment]
Thanks @gchanan . I'm working on an update right now. I pushed some commits because I need to switch to another develop machine. I will notify here when the PR is ready to be reviewed.

**Comment by SsnL** [comment]
Hi @gchanan , I've addressed your comments in new commits. Now:
1. The `int64_t`s (`long`s) in spatial versions stay as `int64_t`s.
2. Corresponding variables in volumetric versions are now `int64_t`s.
3 I propose changing spatial CUDA kernel's `stride` inputs to type `int64_t` because the caller has them in `int64_t`. So this fixes the case that everything but strides fit into `int`s in CUDA spatial case. This is done in a separate commit `changed stride in spatial CUDA kernel to int64_t`.
4. Rebased & merged.

Thanks again for your comments!

**Review by gchanan — CHANGES_REQUESTED** [review]
The python/docs changes all look good.  I'm not planning to look at those again unless you tell me they changed.

I'm still not really confident I can review the C/CUDA changes properly as is; the first commit contains variable renames, ordering changes, and logic changes it ways that don't seem consistent (i.e. I think I caught some mistakes but it's hard to tell with the mass of changes and it's likely I missed some others).  There also seems to be an effort to maintain consistency across C/CUDA implementations, but not across Spatial/Volumetric with types and such; that doesn't seem like the right axis to maintain consistency on; they are different languages, you are optimizing for different things, etc.

Here is how I think an "ideal" review of the C/CUDA would go:
1) First commit(s) contain existing Spatial code with only renames
2) Next commit(s) contain logic/reordering changes to existing Spatial code
3) Next commit(s) contain Volumetric changes.  Volumetric changes are verified by diffing the corresponding Spatial versions (after 2 above) against the Volumetric versions.  The only differences are due to it being Volumetric, i.e. extra for loops, stride calculations.

Does that make sense?

File: torch/lib/THNN/generic/SpatialAdaptiveAveragePooling.c [hunk 8]
```diff
@@ -7,49 +7,51 @@
...
-#pragma omp parallel for private(k)
-  for (k = 0; k < nslices; k++)
+  int64_t d;
+#pragma omp parallel for private(d)
+  for (d = 0; d < sizeD; d++)
   {
     /* loop over output */
-    int64_t i, j;
-    for(i = 0; i < oheight; i++)
+    int oh, ow;
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/lib/THNN/generic/SpatialAdaptiveAveragePooling.c]: Missed that. Sorry.

**Comment by SsnL** [comment]
Thanks @gchanan ! I've restructured the commits and `push --force`. Sorry for causing these inconvenience. In my other PR (adaptive max pool 3d), I tried to follow your advice and structure them better. So that one should be much easier to review :). 

Let me know if there is anything more I need to change!

File: torch/lib/THNN/generic/SpatialAdaptiveAveragePooling.c [hunk 9]
```diff
@@ -89,96 +92,99 @@ void THNN_(SpatialAdaptiveAveragePooling_updateOutput)(
...
     {
-      THNN_(SpatialAdaptiveAveragePooling_updateOutput_frame)(input_data+p*istride_b, output_data+p*nslices*owidth*oheight,
-                                                        nslices,
-                                                        iwidth, iheight,
-                                                        owidth, oheight,
-                                                        istride_w,istride_h,
-                                                        istride_d);
+      THNN_(SpatialAdaptiveAveragePooling_updateOutput_frame)(input_data+b*istrideB, output_data+b*sizeD*osizeH*osizeW,
+                                                        sizeD,
+                                                        isizeH, isizeW,
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/lib/THNN/generic/SpatialAdaptiveAveragePooling.c]: Thanks for pointing that out. Resolved in updated commits!

**Comment by gchanan** [comment]
btw, closing and reopening the PR should trigger travis so you don't have to do an empty commit.

**Comment by SsnL** [comment]
Thanks for the tip @gchanan :)

File: torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+  input_data = THCTensor_(data)(state, input);
+  output_data = THCTensor_(data)(state, output);
+
+  int64_t offsetZ = 0;
+  dim3 threads(32, 8);
+  // each H*W plane is processed by blocksH thread blocks
+  int blocksH = max((int)(16L / totalZ), 1);
+  while (totalZ > 0) {
+    dim3 blocks(totalZ > 65535 ? 65535 : totalZ, blocksH);
```
Comment (gchanan) [inline torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu]: why did you choose to break things up this way (blocks,threads) (where the blocks are 2d), instead of the existing Volumetric pooling (VolumetricAveragePooling, VolumetricDilatedMaxPooling) where it's (grid/blocks) and grid is 3d?

**Review by gchanan — COMMENTED** [review]
Looks good.  Non-CUDA changes are ready; I haven't done much CUDA programming, so I'm less confident about that.  If you could explain my question below it would help.

File: torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu (hunk 10 above)
Comment (SsnL) [reply torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu]: `(blocks,threads)` naming is following the SpatialAdaptiveAveragePooling.
That blocks is 2d is also following SpatialAdaptiveAveragePooling, where each thread does multiple pixels rather than 1 (`blocksH` blocks together process a H*W plane). In VolumetricDilatedMaxPooling, H*W/(32*8) blocks process a plane. 

I'm not sure if one way is better than the other. But I thought following spatial version should be a reasonable thing to do. :)

File: torch/lib/THCUNN/VolumetricAdaptiveAveragePooling.cu [hunk 11]
```diff
@@ -0,0 +1,250 @@
+#include "THCUNN.h"
+#include "THCHalf.h"
+#include "THCDeviceTensor.cuh"
```
[the remark this answers is not in the document]
Comment (SsnL) [reply torch/lib/THCUNN/VolumetricAdaptiveAveragePooling.cu]: Good catch! They are not needed. I forgot to remove them after removing a function call. I updated the PR.

File: torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+    isizeT = input->size[2];
+    isizeH = input->size[3];
+    isizeW = input->size[4];
+
+    osizeT = gradOutput->size[2];
+    osizeH = gradOutput->size[3];
+    osizeW = gradOutput->size[4];
+  }
+
+  // somehow nonatomic is passing all test for volumetric case.
```
Comment (gchanan) [inline torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu]: it should, assuming it is written correctly, right?  My understanding is that nonatomic means it's not using any atomic operations (e.g. atomicAdd); since it's operating over the input pixels this is okay because only one thread is writing to that location in the grad input.  The atomic version is operating over the output pixels; since an input can correspond to multiple outputs it needs to use atomic ops.

**Review by gchanan — APPROVED** [review]
looks good!

File: torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu (hunk 12 above)
Comment (SsnL) [reply torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu]: Yeah, it should. Spatial version said its nonatomic kernel isn't passing the tests and thus disabled it. Either there is something weird going on with spatial kernel, or the author just copied from spatial adaptive max pooling, which also has that line (https://github.com/pytorch/pytorch/blob/master/torch/lib/THCUNN/generic/SpatialAdaptiveMaxPooling.cu#L97). But my volumetric adaptive max pooling also works for the nonatomic case in tests #2782 , so I'm not really sure what's going on.

**Comment by soumith** [comment]
thanks a ton @SsnL !!!

**Comment by SsnL** [comment]
Thanks @soumith ! Glad to contribute!

File: torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu (hunk 12 above)
Comment (fmassa) [reply torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu]: For the record, if I remember properly the `SpatialAdaptiveMaxPooling` didn't indeed pass the tests without atomic operations , so I just used it by default and didn't look back to see why it was the case :-)

Comment (SsnL) [reply torch/lib/THCUNN/generic/VolumetricAdaptiveAveragePooling.cu]: Thanks for clearing things up :)
